README.rst
===================
Twitter API Crawler
===================


.. image:: https://img.shields.io/pypi/v/twitter_api_crawler.svg
        :target: https://pypi.python.org/pypi/twitter_api_crawler

.. image:: https://img.shields.io/travis/Carabinr/twitter_api_crawler.svg
        :target: https://travis-ci.com/Carabinr/twitter_api_crawler

.. image:: https://readthedocs.org/projects/twitter-api-crawler/badge/?version=latest
        :target: https://twitter-api-crawler.readthedocs.io/en/latest/?version=latest
        :alt: Documentation Status




A robust and scalable API client to crawl Twitter API politely and within limits.


* Free software: MIT license
* Documentation: https://twitter-api-crawler.readthedocs.io.


Features
--------

* TODO

Credits
-------

This package was created with Cookiecutter_ and the `audreyr/cookiecutter-pypackage`_ project template.

.. _Cookiecutter: https://github.com/audreyr/cookiecutter
.. _`audreyr/cookiecutter-pypackage`: https://github.com/audreyr/cookiecutter-pypackage


HISTORY.rst
=======
History
=======

0.1.0 (2022-09-04)
------------------

* First release on PyPI.


tests/__init__.py
"""Unit test package for twitter_api_crawler."""


tests/test_api.py
import os
import unittest
from twitter_api_crawler.api import TwitterAPIv1
from twitter_api_crawler.exceptions import Twitter404Exception



class TestTwitterAPIv1(unittest.TestCase):

    def setUp(self) -> None:
        self.api = TwitterAPIv1(
            api_key=f"test_api_key",
            api_key_secret='b',
            access_token='c',
            access_token_secret='d'
        )
        self.real_api = TwitterAPIv1(
            api_key=os.getenv('API_KEY'),
            api_key_secret=os.getenv('API_KEY_SECRET'),
            access_token=os.getenv('ACCESS_TOKEN'),
            access_token_secret=os.getenv('ACCESS_TOKEN_SECRET'),
            cache_requests=False
        )
        self.empty_obj = {
            "id": 1234,
            "entities": {
            }
        }
        self.single_obj= {
            "id": 1234,
            "entities": {
                "user_mentions": [
                    {
                        "screen_name": "twitterapi",
                        "name": "Twitter API",
                        "id": 6253282,
                        "id_str": "6253282",
                        "indices": [11,22]
                    }
                ],
                "hashtags": [
                    {
                        "text": "hashtag",
                        "indices": [
                            23,
                            31
                        ]
                    }
                ],
                "symbols": [
                    {
                        "text": "TWTR",
                        "indices": [
                            5,
                            10
                        ]
                    }
                ],
                "url": {
                    "urls": [
                        {
                            "url": "http:\/\/t.co\/p5dOtmnZyu",
                            "expanded_url": "http:\/\/dev.twitter.com",
                            "display_url": "dev.twitter.com",
                            "indices": [32,54]
                        }
                    ]
                }
            }
        }


    def test_lookup_group_of_missing_users_with_404(self):
        # screen_name='Bowtiedspiderc1,camperkenny,GardenOfMyopia,petr_bz,Wealth_Theory,wvaeu'
        screen_name='Bowtiedspiderc1,camperkenny'

        with self.assertRaises(Twitter404Exception):
            self.real_api.lookup_users(screen_name)

    def test_get_mentions_empty(self):
        output = self.api.get_mentions(self.empty_obj)
        self.assertEqual(len(output), 0)

    def test_get_mentions_single(self):
        output = self.api.get_mentions(self.single_obj)
        self.assertEqual(len(output), 1)

    def test_get_mentions_multiple(self):
        self.single_obj['entities']['user_mentions'].append(
            {
                "screen_name": "twitterapi2",
                "name": "Twitter API2",
                "id": 62532822,
                "id_str": "62532822",
                "indices": [112, 122]
            }
        )

        output = self.api.get_mentions(self.single_obj)
        self.assertEqual(len(output), 2)

    def test_get_hashtags_empty(self):
        output = self.api.get_hashtags(self.empty_obj)
        self.assertEqual(len(output), 0)

    def test_get_hashtags_single(self):
        output = self.api.get_hashtags(self.single_obj)
        self.assertEqual(len(output), 1)

    def test_get_hashtags_multiple(self):
        self.single_obj['entities']['hashtags'].append({"text": "hashtag2","indices": [23,31]})
        output = self.api.get_hashtags(self.single_obj)
        self.assertEqual(len(output), 2)

    def test_get_urls_empty(self):
        output = self.api.get_urls(self.empty_obj)
        self.assertEqual(len(output), 0)

    def test_get_urls_single(self):
        output = self.api.get_urls(self.single_obj)
        self.assertEqual(len(output), 1)

    def test_get_urls_multiple(self):
        urlobj = {
            "url": "http:\/\/t.co\/p5dOtmnZyu2",
            "expanded_url": "http:\/\/dev.twitter.com2",
            "display_url": "dev.twitter.com2",
            "indices": [
                132,
                154
            ]
        }

        self.single_obj['entities']['url']['urls'].append(urlobj)
        output = self.api.get_urls(self.single_obj)
        self.assertEqual(len(output), 2)


    def test_get_mentions_from_text_one_result(self):
        description = "Hi, I'm bob. I work for @heysamtexas"
        out = self.api.get_mentions_from_text(description)
        self.assertEqual(len(out), 1)
        self.assertEqual(out[0], 'heysamtexas')

    def test_get_mentions_from_text_multiple_results(self):
        description = "Hi, I'm bob. I work for @heysamtexas formerly @king"
        out = self.api.get_mentions_from_text(description)
        self.assertEqual(len(out), 2)
        self.assertEqual(out[0], 'heysamtexas')

    def test_get_mentions_from_text_empty_string(self):
        description = ""
        out = self.api.get_mentions_from_text(description)
        self.assertEqual(len(out), 0)

    def test_get_mentions_from_text_no_mentions(self):
        description = "This is a description without a mention"
        out = self.api.get_mentions_from_text(description)
        self.assertEqual(len(out), 0)

    def test_get_mentions_from_text_with_email(self):
        description = "This is a description without a mention. bob@idiot.com"
        out = self.api.get_mentions_from_text(description)
        self.assertEqual(len(out), 0)

    def test_get_hashtags_from_text_empty(self):
        description = ''
        out = self.api.get_hashtags_from_text(description)
        self.assertEqual(len(out), 0)

    def test_get_hashtags_from_text_one_hashtag(self):
        description = 'This is my #hashtag'
        out = self.api.get_hashtags_from_text(description)
        self.assertEqual(len(out), 1)
        self.assertEqual(out[0], 'hashtag')

    def test_get_hashtags_from_text_multiple_hashtags(self):
        description = 'This is my #hashtag. There is another one #here.'
        out = self.api.get_hashtags_from_text(description)
        self.assertEqual(len(out), 2)
        self.assertEqual(out[1], 'here')

    def test_get_hashtags_from_text_multiple_hashtags_tricky(self):
        description = 'This is my #hashtag. There is another one #here.#here2'
        out = self.api.get_hashtags_from_text(description)
        self.assertEqual(len(out), 3)
        self.assertEqual(out[2], 'here2')

    def test_get_hashtags_from_text_with_tricky_url(self):
        description = 'This is my #hashtag. https://www.bob.com/#bob'
        out = self.api.get_hashtags_from_text(description)
        self.assertEqual(len(out), 1)
        self.assertEqual(out[0], 'hashtag')


tests/test_crawler.py
import time
import unittest
from twitter_api_crawler.api import TwitterAPIv1
from twitter_api_crawler.crawler import TwitterAPIv1_Crawler


class TestTwitterAPIv1_Crawler(unittest.TestCase):

    def setUp(self) -> None:
        self.crawler = TwitterAPIv1_Crawler()

    def tearDown(self) -> None:
        self.crawler = None

    def create_api(self, key:str):
        self.crawler.create_api(
            key=key,
            api_key=f"{key}_api_key",
            api_key_secret='b',
            access_token='c',
            access_token_secret='d'
        )


    def test_create_empty(self):
        self.assertEqual(len(self.crawler.apis), 0)

    def test_create_api(self):
        self.create_api('boom')
        self.assertEqual(len(self.crawler.apis), 1)
        self.assertEqual(self.crawler.current_key, 'boom')

    def test_create_api_duplicate_key(self):

        with self.assertRaises(Exception):
            self.create_api('boom')
            self.create_api('boom')

    def test_create_api_does_not_change_current_key(self):
        self.create_api('bob')
        self.create_api('bob2')

        self.assertEqual(self.crawler.current_key, 'bob')

    def test_get_api(self):
        self.create_api('boom')
        cred = self.crawler.get_api('boom')
        self.assertIsInstance(cred, TwitterAPIv1)

    def test_get_api_missing_key(self):
        self.create_api('boom')
        cred = self.crawler.get_api('boom2')
        self.assertIsNone(cred)

    def test_delete_api(self):
        self.create_api('boom')
        cred = self.crawler.get_api('boom')
        self.assertIsInstance(cred, TwitterAPIv1)

        self.crawler.delete_api('boom')
        self.assertEqual(len(self.crawler.apis), 0)

    def test_update_api_missing_key(self):
        self.create_api('boom')

        with self.assertRaises(Exception):
            self.crawler.update_api(
                key='boom2',
                api_key='a',
                api_key_secret='b',
                access_token='c',
                access_token_secret='d'
            )

    def test_update_api(self):
        self.create_api('boom')

        self.crawler.update_api(
            key='boom',
            api_key='1',
            api_key_secret='2',
            access_token='3',
            access_token_secret='4'
        )
        self.assertEqual(len(self.crawler.apis), 1)

    def test_get_cursor_when_empty(self):
        cursor = self.crawler.get_cursor('bob')
        self.assertEqual(cursor, "-1")

    def test_get_cursor_when_present(self):
        self.crawler.set_cursor('bob', '9999')
        cursor = self.crawler.get_cursor('bob')
        self.assertEqual(cursor, "9999")

    # def test_get_next_key(self):
    #     self.create_api('bob')
    #     self.assertEqual(self.crawler.get_next_key(), 'bob')
    #
    #     self.create_api('bob2')
    #     self.assertEqual(self.crawler.current_key, 'bob')
    #
    #     self.assertEqual(self.crawler.get_next_key(), 'bob2')
    #
    # def test_get_next_key_when_nos(self):
    #     self.assertIsNone(self.crawler.get_next_key())

    def test_fetch_api_when_empty(self):
        self.assertIsNone(self.crawler.fetch_api())

    def test_next_api_advances_next_key(self):
        self.create_api('boom')
        first_key = self.crawler.current_key
        self.create_api('boom2')
        self.create_api('boom3')
        self.create_api('boom4')
        self.create_api('boom5')
        self.assertEqual(self.crawler.current_key, first_key)

        # gets current api without iterating to next
        cred1 = self.crawler.fetch_api()
        api_key = cred1.auth.client.client_key
        self.assertEqual(api_key, 'boom_api_key')

        self.crawler.pause_current_api(60)
        cred2 = self.crawler.next_api()
        api_key = cred2.auth.client.client_key
        self.assertNotEqual(self.crawler.current_key, first_key)

    def test_next_api_loops_around(self):
        self.create_api('boom')
        self.create_api('boom2')
        self.create_api('boom3')

        # test that it loops around
        looped = self.crawler.next_api()
        looped = self.crawler.next_api()
        looped = self.crawler.next_api()
        api_key = looped.auth.client.client_key
        self.assertEqual(api_key, 'boom_api_key')
        self.assertEqual(self.crawler.current_key, 'boom')

    def test_next_api_is_none_when_empty(self):
        cred1 = self.crawler.next_api()
        self.assertIsNone(cred1)

    def test_pause_current_api(self):
        self.create_api('boom')
        self.crawler.pause_current_api(15*60)

        self.assertTrue(self.crawler.fetch_api().is_asleep())

    def test_pause_current_api_short_time(self):
        self.create_api('boom')
        self.crawler.pause_current_api(2)
        self.assertTrue(self.crawler.fetch_api().is_asleep())

        time.sleep(3)

        self.assertFalse(self.crawler.fetch_api().is_asleep())

    def test_next_api__all_apis_paused_is_none(self):
        self.create_api('boom')
        self.create_api('boom2')
        self.create_api('boom3')

        self.crawler.pause_current_api(600)
        out1 = self.crawler.next_api()

        self.crawler.pause_current_api(600)
        out2 = self.crawler.next_api()

        self.crawler.pause_current_api(600)
        out3 = self.crawler.next_api()

        self.assertIsNone(self.crawler.next_api())

    def test_next_api__no_apis_created(self):
        self.create_api('boom')
        self.crawler.pause_current_api(600)
        self.crawler.next_api()
        self.assertIsNone(self.crawler.next_api())

    def test_get_all_following_no_api(self):
        self.create_api('boom')
        self.crawler.pause_current_api(60)
        payload = self.crawler.get_all_following('heysamtexas')
        self.assertEqual(len(payload['users']), 0)
        self.assertFalse(payload['completed'])
        self.assertEqual(payload['cursor'], "-1")
        print(payload)


twitter_api_crawler/crawler.py
from typing import Union, Dict
from loguru import logger
from .exceptions import TwitterNoAvailableAPIs, Twitter429Exception
from .api import TwitterAPIv1
from requests_oauthlib import OAuth1


class TwitterAPIv1_Crawler(object):
    apis = {}
    current_key = ''
    cursors = {}

    def __init__(self):
        self.apis = {}
        self.current_key = ''
        self.cursors = {}

    def create_api(self, key:str, api_key:str, api_key_secret:str, access_token:str, access_token_secret:str):
        """

        @param key:
        @param api_key:
        @param api_key_secret:
        @param access_token:
        @param access_token_secret:
        @return:
        """
        if key in self.apis:
            raise Exception('Duplicate Key in your credentials')

        if self.current_key == '':
            self.current_key = key

        self.apis[key] = TwitterAPIv1(api_key, api_key_secret, access_token, access_token_secret)


    def get_api(self, key:str):
        return self.apis.get(key, None)

    def delete_api(self, key:str) -> None:
        if key in self.apis:
            del self.apis[key]


    def update_api(self, key:str, api_key:str, api_key_secret:str, access_token:str, access_token_secret:str):
        if key not in self.apis:
            raise Exception('That key does not exist')

        self.apis[key] = OAuth1(api_key, api_key_secret, access_token, access_token_secret)

    def fetch_api(self) -> Union[TwitterAPIv1, None]:
        if len(self.apis) == 0:
            return None

        return self.apis[self.current_key]


    def next_api(self) -> Union[TwitterAPIv1, None]:
        """
        Grab the next available API client from our list that is not asleep.

        Returns:
            None (if none are available) or a TwitterAPI client
        """
        for k in list(self.apis.keys()):
            if not self.apis[k].is_asleep():
                self.current_key = k
                return self.apis[k]

        self.current_key = None
        return None

    def pause_current_api(self, secs:int) -> None:
        """
        Calling this method will set a sleep property on the TwitterAPI object. If it is set (and still valid), then
        it will not be included when rotating API clients.

        Args:
            secs: the number of seconds this api client should sleep

        Returns:
            None
        """
        if self.current_key:
            self.apis[self.current_key].sleep(secs)

    def set_cursor(self, username:str, cursor:str) -> None:
        self.cursors[username] = cursor

    def get_cursor(self, username:str) -> str:
        return self.cursors.get(username, '-1')

    def get_all_following(self, username: str, cursor: str = "-1") -> Dict:
        """
        Given a Username (eg Twitter screen_name) and optional Twitter "cursor", try to crawl as many followed
        accounts as possible with the assigned Twitter Objects.

        If API Rate limits are hit (HTTP 429), then it will mark that particular API client as asleep and rotate
        to another client in the list.

        Once it runs out of available API clients, it will exit the loop and return the users it has up to that point.
        It will also return the last known cursor so you can save your process.

        Args:
            username (str): username (eg. screen_name) of a known Twitter user
            cursor (int): An integer that is used for pagination with the Twitter API

        Returns:
            Python dict containing username(str), users(list of users crawled), cursor (int), and completed(boolean)
        """
        has_more = True
        output = []
        completed = False

        while has_more:
            api = self.next_api()

            if api is None:
                has_more = False
                break

            try:
                results = api.get_following(username, cursor)
            except Twitter429Exception:
                logger.info('Got 429: Marking API Client Key as unavailable')
                self.pause_current_api(15*60)
                continue

            users = results.get('users', [])
            cursor = results.get('next_cursor')

            if len(users) > 0:
                output.append(users)
            else:
                completed = True
                has_more = False

        flat_users = [item for sublist in output for item in sublist]

        payload = {
            'username': username,
            'users': flat_users,
            'cursor': cursor,
            'completed': completed
        }

        return payload

    def get_following(self, username: str, cursor: int = -1) -> Dict:
        """
        Given a Username (eg Twitter screen_name) and optional Twitter "cursor", try to crawl as many followed
        accounts as possible with the assigned Twitter Objects.

        If API Rate limits are hit (HTTP 429), then it will mark that particular API client as asleep and rotate
        to another client in the list.

        Once it runs out of available API clients, it will exit the loop and return the users it has up to that point.
        It will also return the last known cursor so you can save your process.

        Args:
            username (str): username (eg. screen_name) of a known Twitter user
            cursor (int): An integer that is used for pagination with the Twitter API

        Returns:
            Python dict containing username(str), users(list of users crawled), cursor (int), and completed(boolean)

        Raises:
            Twitter429Exception
        """
        output = []
        completed = False

        api = self.next_api()

        if api is None:
            raise TwitterNoAvailableAPIs()

        try:
            results = api.get_following(username, cursor)
        except Twitter429Exception:
            logger.info('Got 429: Marking API Client Key as unavailable for 15 minutes')
            self.pause_current_api(15 * 60)
            raise Twitter429Exception()

        users = results.get('users', [])
        cursor = results.get('next_cursor')

        if len(users) > 0:
            output.append(users)

        flat_users = [item for sublist in output for item in sublist]

        payload = {
            'username': username,
            'users': flat_users,
            'cursor': cursor,
            'completed': cursor == 0
        }

        return payload


twitter_api_crawler/api.py
import requests
from requests_oauthlib import OAuth1
from typing import Dict, List
import datetime
from loguru import logger
import re
from requests_cache import CachedSession
from .exceptions import Twitter429Exception, Twitter404Exception


class TwitterAPIv1(object):

    auth = None
    sleep_until = None
    cache_requests = False

    def __init__(self,
                 api_key: str,
                 api_key_secret: str,
                 access_token: str,
                 access_token_secret: str,
                 cache_requests: bool = False):
        self.auth = OAuth1(api_key, api_key_secret, access_token, access_token_secret)
        self.sleep_until = None
        self.cache_requests = cache_requests

    def sleep(self, secs):
        self.sleep_until = datetime.datetime.now(datetime.timezone.utc) + datetime.timedelta(seconds=secs)

    def wakeup(self):
        self.sleep_until = None

    def is_asleep(self):
        now = datetime.datetime.now(datetime.timezone.utc)

        if self.sleep_until is None:
            return False

        if now > self.sleep_until:
            self.wakeup()
            return False

        return True

    def _call(self, method:str, url:str, params: Dict = None, data: Dict = None):
        """
        Secret method that dispatches get, post, put, and delete requests into the Session() module
        Args:
            method (str): get, post, put, or delete all lower case. The HTTP method
            url (str): The Full URL of the destination
            params (dict): an optional dictionary of parameters passed into the HTTP request

        Returns:
            A dict of the Twitter API response body

        Raises:
            Twitter429Exception: when API response with HTTP 429 (rate-limit)
            Twitter404Exception: when the API response returns a 404. Often because the screen_names are no
                                 longer accounts

        """
        session = CachedSession() if self.cache_requests else requests.session()

        r = getattr(session, method)(url=url, auth=self.auth, params=params, data=data)

        logger.debug(f'Fetched: {r.url}')
        logger.debug(f'Got HTTP Response: {r.status_code}')

        if r.status_code == 400:
            logger.debug(r.content)

        if r.status_code == 429:
            raise Twitter429Exception()

        if r.status_code == 404:
            raise Twitter404Exception()

        return r

    def _get(self, url:str, params: Dict = None):
        return self._call("get", url, params)

    def _post(self, url:str, params: Dict = None, data: Dict = None):
        return self._call("post", url, params, data)

    def lookup_users(self, screen_name: str) -> List[Dict]:
        """
        Lookup a user in the Twitter API. Up to 100 usernames in CSV string may be submitted

        Args:
            screen_name: CSV string of Twitter accounts (up to 100)

        Returns:
            A dict object API response

        """
        logger.debug(f"Looking up {screen_name}")
        url = 'https://api.twitter.com/1.1/users/lookup.json'
        data = {'screen_name': screen_name}
        r = self._post(url, params={}, data=data)

        return r.json()

    def get_followers(self, screen_name:str, cursor:int = -1) -> Dict:

        url = 'https://api.twitter.com/1.1/followers/list.json'
        completed = False
        output = []

        r = self._post(url, params={'count': 200, 'cursor': cursor, 'screen_name': screen_name})
        results = r.json()

        if len(results['users']) > 0:
            output.append(results['users'])
            cursor = results['next_cursor']
        else:
            completed = True

        return {
            'users': [item for sublist in output for item in sublist],
            'cursor': cursor,
            'completed': completed
        }

    def get_all_following(self, screen_name):
        cursor = -1
        has_more = True
        output = []
        completed = False

        while has_more:
            results = self.get_following(screen_name, cursor)
            users = results.get('users', [])

            if results.status_code == 429:
                logger.info('Got 429: Bailing out and saving cursor')
                has_more = False

            elif len(users) > 0:
                output.append(users)
                cursor = results['next_cursor']
            else:
                completed = True
                has_more = False

        payload = {
            'users': [item for sublist in output for item in sublist],
            'cursor': cursor,
            'completed': completed
        }

        return payload

    def get_following(self, screen_name: str, cursor: int = -1) -> Dict:
        """
        Get the users that screen_name is following.

        Args:
            screen_name: Twitter account name
            cursor: The current position / offset of results

        Returns:
            Twitter API response body
        """
        url = 'https://api.twitter.com/1.1/friends/list.json'

        params = {
            'count': 200,
            'cursor': cursor,
            'screen_name': screen_name
        }
        r = self._get(url, params)

        return r.json()

    @staticmethod
    def get_mentions(response:Dict) -> List[str]:
        """
        Given an API response status object, return the mentions as a list of strings or empty list

        Args:
            response: API User response object

        Returns:
            List of strings or empty list
        """
        mentions = response.get('entities',{}).get('user_mentions',[])
        usernames = [name.get('screen_name') for name in mentions]
        return usernames

    @staticmethod
    def get_mentions_from_text(text:str) -> List[str]:
        result = re.findall("(^|[^@\w])@(\w{1,15})", text)
        return [_[1] for _ in result]

    @staticmethod
    def get_hashtags_from_text(text:str) -> List[str]:
        rem_url = re.sub(r"https?:[^\s]+",'', text)
        rem_periods = rem_url.replace('.',' ')
        pattern = '(?:(?<=\s)|^)#(\w*[A-Za-z\d\-]{2,60}\w*)'
        return [hashtag_match.group(1) for hashtag_match in re.finditer(pattern, rem_periods)]
    @staticmethod
    def get_hashtags(u:Dict) -> List[str]:
        """
        Given an API response user object, return the hashtags as a list of strings or empty list

        @param u: API User response object
        @return: List of strings or empty list
        """
        mentions = u.get('entities',{}).get('hashtags',[])
        hashtags = [name.get('text') for name in mentions]
        return hashtags

    @staticmethod
    def get_urls(u:Dict) -> List[str]:
        """
        Given an API response user object, return the urls as a list of strings or empty list

        @param u: API User response object
        @return: List of strings or empty list
        """
        url_objs = u.get('entities',{}).get('url',{}).get('urls',[])

        urls = [name.get('expanded_url') for name in url_objs if name.get('expanded_url')]

        if u.get('url'):
            urls.append(u.get('url'))

        return urls
